File: apps/api/app/schemas/shared_income.py
# Request/response schemas for a group's shared income.

from datetime import date as date_type
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator, model_validator

from app.models.group_money_settings import SplitMethod
from app.models.income_entry import IncomeCategory
from app.models.shared_income import IncomeDestination
from app.schemas.base import RequestBase, validate_supported_currency, validate_user_pickable_income_category


# One participant's line in a split. `figure` is the figure the chosen method needs and nothing else:
# ignored by `equal`, an amount for `exact`, a weight for `shares`, a percentage for `percentage`. One
# field rather than three, because exactly one is ever meaningful and three would allow a body where
# two are set and neither is authoritative.
class SharedIncomeSplitInput(RequestBase):
    member_id: int = Field(description="Seat taking a share of this income.")
    figure: Decimal | None = Field(default=None, description="Amount, weight or percentage, per the split method.", max_digits=18, decimal_places=2)


# Body for POST /groups/{group_id}/income.
#
# `destination` says WHERE the money ended up and `received_by_member_id` says WHO holds it; the two
# are one question answered on two branches rather than two independent fields:
#   * `joint` — it landed in a shared account a pot holds, so the pot's owners received it in their own
#     proportions. `paid_to_account_id` is required (a CHECK enforces the shape and the service checks
#     the account is really this group's), and naming a recipient is refused, because saying one member
#     received joint money asserts something the ownership ledger contradicts.
#   * `distributed` — it reached one person. `received_by_member_id` is required; `paid_to_account_id`
#     is optional and, when given, must be that member's own private account.
#
# `source_investment_id` is the co-owned asset the income came from. It drives the DEFAULT split the
# form pre-fills (F1) and is stored as the row's label; it never changes what the API computes, which
# is always the split the request states.
class SharedIncomeCreate(RequestBase):
    date: date_type = Field(description="Income date.")
    amount: Decimal = Field(description="Full amount received, before any split.", gt=0, max_digits=18, decimal_places=2)
    currency: str = Field(description="Currency (ISO 4217); also the balance bucket this lands in.", max_length=3)
    category: IncomeCategory | None = Field(default=None, description="Income category.")
    notes: str | None = Field(default=None, description="Optional notes.", max_length=500)
    split_method: SplitMethod = Field(description="How the total is divided between the participants.")
    splits: list[SharedIncomeSplitInput] = Field(description="Who takes a share, and their figure for the chosen method.", min_length=1)
    destination: IncomeDestination = Field(description="Whether the money stays joint in a pot or is distributed to one person.")
    source_investment_id: int | None = Field(default=None, description="Co-owned asset this came from; must be one this group's pots hold.")
    received_by_member_id: int | None = Field(default=None, description="Seat the money reached; required for distributed, refused for joint.")
    paid_to_account_id: int | None = Field(default=None, description="Account the money arrived in: one a pot holds, or the recipient's own.")

    _validate_currency = field_validator("currency")(validate_supported_currency)
    _validate_category = field_validator("category")(validate_user_pickable_income_category)

    # One seat appears at most once. A second line for the same member is two opinions about one
    # person's share, and the DB's UNIQUE would reject it as an opaque integrity error.
    @model_validator(mode="after")
    def validate_distinct_participants(self) -> "SharedIncomeCreate":
        member_ids = [split.member_id for split in self.splits]
        if len(set(member_ids)) != len(member_ids):
            raise ValueError("Each person can appear only once in a split.")
        return self


# Body for PUT /groups/{group_id}/income/{income_id}. A FULL replacement rather than a partial update,
# for the reason the expense mirror is: the amount, the method and the participants are one
# interlocking statement — changing the amount without restating the split would leave exact figures
# that no longer add up to it, and there is no honest way to infer what the user meant.
class SharedIncomeUpdate(SharedIncomeCreate):
    pass


# One member's position in one piece of shared income: what they are entitled to and what reached them.
class SharedIncomeSplitResponse(BaseModel):
    model_config = {"from_attributes": True}

    member_id: int = Field(description="Seat this position belongs to.")
    display_name: str = Field(description="How that person is shown in the group.")
    amount: Decimal = Field(description="What this member is entitled to — their share of the income.")
    received_amount: Decimal = Field(description="What actually reached this member.")
    is_self: bool = Field(description="Whether this is the requesting user's own seat.")


# Response for the shared-income list and for POST / PUT.
#
# `received_by_member_id` / `received_by_display_name` are DERIVED rather than stored, and are null
# whenever the money landed in a SHARED account — the pot's owners received it, in their own
# proportions, and a stored receiver column could not express that, which is why there is not one.
#
# Decided by the DESTINATION rather than by the shape of the splits, for the same reason the expense
# mirror is decided by the funding: a pot with exactly ONE owner has that owner receiving the whole
# amount, which is indistinguishable from one person collecting it — and reporting them as the
# recipient would say somebody took the money personally when it went into the joint account.
class SharedIncomeResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int = Field(description="Shared income id.")
    group_id: int = Field(description="Group that shares it.")
    date: date_type = Field(description="Income date.")
    amount: Decimal = Field(description="Full amount received, before any split.")
    currency: str = Field(description="Currency (ISO 4217).")
    converted_amount: Decimal | None = Field(default=None, description="Full amount in the requested display currency.")
    category: IncomeCategory | None = Field(default=None, description="Income category.")
    notes: str | None = Field(default=None, description="Optional notes.")
    split_method: SplitMethod = Field(description="How the total was divided.")
    destination: IncomeDestination = Field(description="Whether it stayed joint in a pot or was distributed.")
    source_investment_id: int | None = Field(default=None, description="Co-owned asset it came from.")
    source_investment_name: str | None = Field(
        default=None, description="That asset's name; null when it is gone or the caller cannot see the pot holding it."
    )
    paid_to_account_id: int | None = Field(default=None, description="Account the money arrived in.")
    paid_to_account_name: str | None = Field(default=None, description="That account's name, so the row reads without a second request.")
    received_by_member_id: int | None = Field(default=None, description="Seat the money reached; null when it landed in a shared account.")
    received_by_display_name: str | None = Field(default=None, description="That person's name; null when it landed in a shared account.")
    my_share: Decimal | None = Field(default=None, description="The requesting user's own share; null when they take none.")
    splits: list[SharedIncomeSplitResponse] = Field(description="Every member's position in this income.")
    created_at: datetime = Field(description="Creation timestamp.")
    updated_at: datetime = Field(description="Last update timestamp.")

File: apps/web/lib/api/shared-income.ts
/*
 * Server-only types + mappers for a group's shared income — the income half of the flow half. Reads go
 * through `getSharedIncome`; mutations are server actions in
 * `app/(protected)/shared/shared-income-actions.ts`.
 *
 * Two shapes of the API's honesty survive into these types unchanged, and the UI must respect both.
 * They are the mirror of the shared-expense pair, with the two sides swapped:
 *
 *   * A member's position in one income row is TWO figures, not one: what they are entitled to
 *     (`amount`) and what actually reached them (`receivedAmount`). Their balance is the difference,
 *     and both columns sum to the row's total — which is what makes the group's balances sum to zero.
 *     Never render one as if it were the other.
 *   * `receivedByMemberId` is DERIVED from the destination and is null whenever the money landed in a
 *     SHARED account, including a pot with exactly one owner. There is no receiver column precisely
 *     because joint money reaches that pot's owners in their own proportions, so "X received it" must
 *     never appear on a joint row however its ownership happens to be divided.
 *
 * Money figures stay STRINGS end to end, as everywhere else in the app: they are NUMERIC(18,2)
 * server-side, and parsing them into a JS number would round exactly the digits the balance depends on.
 */

import 'server-only';

import { authenticatedFetch } from '@/lib/authenticated-fetch';
import type { SplitMethod } from '@/lib/constants/shared-expenses';
import type { IncomeDestination } from '@/lib/constants/shared-income';

// --- Raw types (API JSON shape, snake_case) ---

interface SharedIncomeSplitRaw {
  member_id: number;
  display_name: string;
  amount: string;
  received_amount: string;
  is_self: boolean;
}

interface SharedIncomeRaw {
  id: number;
  group_id: number;
  date: string;
  amount: string;
  currency: string;
  converted_amount: string | null;
  category: string | null;
  notes: string | null;
  split_method: SplitMethod;
  destination: IncomeDestination;
  source_investment_id: number | null;
  source_investment_name: string | null;
  paid_to_account_id: number | null;
  paid_to_account_name: string | null;
  received_by_member_id: number | null;
  received_by_display_name: string | null;
  my_share: string | null;
  splits: SharedIncomeSplitRaw[];
  created_at: string;
  updated_at: string;
}

// --- Frontend types (camelCase) ---

export interface SharedIncomeSplit {
  memberId: number;
  displayName: string;
  // What this member is entitled to — their share, and the figure that becomes their own income.
  amount: string;
  // What actually reached this member. Zero for somebody still waiting for theirs, the whole amount
  // for a single collector; for a joint row every owner receives their ownership proportion.
  receivedAmount: string;
  isSelf: boolean;
}

export interface SharedIncome {
  id: number;
  groupId: number;
  date: string;
  amount: string;
  currency: string;
  convertedAmount: string | null;
  category: string | null;
  notes: string | null;
  splitMethod: SplitMethod;
  destination: IncomeDestination;
  // The co-owned asset it came from, which is what seeds the default split (F1).
  sourceInvestmentId: number | null;
  /*
   * That asset's name. Null in two legitimate cases rather than one: the asset is gone (the FK is ON
   * DELETE SET NULL, because the money arrived whatever later happened to the asset), or it sits in a
   * pot this viewer is not permitted to see.
   */
  sourceInvestmentName: string | null;
  paidToAccountId: number | null;
  // Denormalized so a row reads without a second request, and so it still names the account after it
  // has been archived — the same reason a card settlement carries its account's name.
  paidToAccountName: string | null;
  /*
   * Who holds it, DERIVED from the destination rather than stored. Null whenever a shared account
   * received it: the pot's owners did, in their own proportions, which no single column could say.
   */
  receivedByMemberId: number | null;
  receivedByDisplayName: string | null;
  // The viewer's own share, or null when they are entitled to none of this row.
  myShare: string | null;
  splits: SharedIncomeSplit[];
  createdAt: string;
  updatedAt: string;
}

// --- Mappers ---

function mapSplit(raw: SharedIncomeSplitRaw): SharedIncomeSplit {
  return {
    memberId: raw.member_id,
    displayName: raw.display_name,
    amount: raw.amount,
    receivedAmount: raw.received_amount,
    isSelf: raw.is_self,
  };
}

function mapSharedIncome(raw: SharedIncomeRaw): SharedIncome {
  return {
    id: raw.id,
    groupId: raw.group_id,
    date: raw.date,
    amount: raw.amount,
    currency: raw.currency,
    convertedAmount: raw.converted_amount,
    category: raw.category,
    notes: raw.notes,
    splitMethod: raw.split_method,
    destination: raw.destination,
    sourceInvestmentId: raw.source_investment_id,
    sourceInvestmentName: raw.source_investment_name,
    paidToAccountId: raw.paid_to_account_id,
    paidToAccountName: raw.paid_to_account_name,
    receivedByMemberId: raw.received_by_member_id,
    receivedByDisplayName: raw.received_by_display_name,
    myShare: raw.my_share,
    splits: raw.splits.map(mapSplit),
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

// --- API functions ---

/*
 * A group's shared income, newest first, each row with every member's position in it.
 *
 * Returns null for a group that does not exist OR that the caller is not a member of — the API
 * answers 404 for both, so the hub renders notFound() either way and an id cannot be probed. The whole
 * list comes back unpaginated, which is what the API offers; the hub pages it client-side.
 */
export async function getSharedIncome(groupId: number): Promise<SharedIncome[] | null> {
  const res = await authenticatedFetch(`/groups/${groupId}/income`, { method: 'GET' });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error('Failed to fetch shared income');
  const raw: SharedIncomeRaw[] = await res.json();
  return raw.map(mapSharedIncome);
}
